# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/GC.cpp
# Commit: e15f1a6f5a38
# Full Hash: e15f1a6f5a38b276216c77f2827945b6f5ec4ea9
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-06 09:33:16
# Regressor Bug: 1689413
# File Overlap Count: 1
# Description:
#   Bug 1689413 part 9 - Move proto from ObjectGroup to BaseShape. r=jonco
#   
#   The ObjectGroup still stores the proto as well but we don't really use it anymore.
#   This is simpler than removing the group's proto now because that would result in
#   "empty" groups and the group table then doesn't make sense anymore.
# ==============================================================================

diff -r 95fa0ec36f7a -r e15f1a6f5a38 js/src/gc/GC.cpp
--- a/js/src/gc/GC.cpp	Fri Mar 05 19:10:08 2021 +0000
+++ b/js/src/gc/GC.cpp	Fri Mar 05 19:10:08 2021 +0000
@@ -8044,14 +8044,14 @@
        !group.done(); group.next()) {
     // Replace placeholder object prototypes with the correct prototype in
     // the target realm.
-    TaggedProto proto(group->proto());
+    TaggedProto proto(group->protoDeprecated());
     if (proto.isObject()) {
       JSObject* obj = proto.toObject();
       if (GlobalObject::isOffThreadPrototypePlaceholder(obj)) {
         JSObject* targetProto =
             global->getPrototypeForOffThreadPlaceholder(obj);
         MOZ_ASSERT(targetProto->isDelegate());
-        group->setProtoUnchecked(TaggedProto(targetProto));
+        group->setProtoUncheckedDeprecated(TaggedProto(targetProto));
       }
     }
   }
@@ -8059,6 +8059,19 @@
   for (auto baseShape = source->zone()->cellIterUnsafe<BaseShape>();
        !baseShape.done(); baseShape.next()) {
     baseShape->setRealmForMergeRealms(target);
+
+    // Replace placeholder object prototypes with the correct prototype in
+    // the target realm.
+    TaggedProto proto = baseShape->proto();
+    if (proto.isObject()) {
+      JSObject* obj = proto.toObject();
+      if (GlobalObject::isOffThreadPrototypePlaceholder(obj)) {
+        JSObject* targetProto =
+            global->getPrototypeForOffThreadPlaceholder(obj);
+        MOZ_ASSERT(targetProto->isDelegate());
+        baseShape->setProtoForMergeRealms(TaggedProto(targetProto));
+      }
+    }
   }
 
   // Fixup zone pointers in source's zone to refer to target's zone.